File: backend/InterviewPractice.Application/Reports/ReportService.cs
using InterviewPractice.Application.Common.Interfaces;
using InterviewPractice.Application.Reports.Dtos;
using InterviewPractice.Domain.Enums;
using Microsoft.EntityFrameworkCore;

namespace InterviewPractice.Application.Reports;

public class ReportService : IReportService
{
    private readonly IApplicationDbContext _dbContext;

    public ReportService(IApplicationDbContext dbContext)
    {
        _dbContext = dbContext;
    }

    public async Task<ReportOverviewDto> GetOverviewAsync(
        CancellationToken cancellationToken = default)
    {
        var totalCandidates = await _dbContext.CandidateProfiles
            .AsNoTracking()
            .CountAsync(cancellationToken);

        var totalInterviews = await _dbContext.Interviews
            .AsNoTracking()
            .CountAsync(cancellationToken);

        var completedInterviews = await _dbContext.Interviews
            .AsNoTracking()
            .CountAsync(
                x => x.Status == InterviewStatus.Completed,
                cancellationToken);

        var scheduledInterviews = await _dbContext.Interviews
            .AsNoTracking()
            .CountAsync(
                x => x.Status == InterviewStatus.Scheduled,
                cancellationToken);

        var averageScore = await _dbContext.Feedbacks
            .AsNoTracking()
            .Select(x => (double?)x.OverallScore)
            .AverageAsync(cancellationToken);

        return new ReportOverviewDto
        {
            TotalCandidates = totalCandidates,
            TotalInterviews = totalInterviews,
            CompletedInterviews = completedInterviews,
            ScheduledInterviews = scheduledInterviews,
            AverageScore = averageScore
        };
    }

    public async Task<CandidateProgressDto?> GetCandidateProgressAsync(
        Guid candidateId,
        CancellationToken cancellationToken = default)
    {
        var candidate = await _dbContext.CandidateProfiles
            .AsNoTracking()
            .Where(x => x.Id == candidateId)
            .Select(x => new
            {
                x.Id,
                FirstName = x.User.FirstName,
                LastName = x.User.LastName
            })
            .FirstOrDefaultAsync(cancellationToken);

        if (candidate is null)
        {
            return null;
        }

        var interviews = await _dbContext.Interviews
            .AsNoTracking()
            .Where(x => x.CandidateId == candidateId)
            .OrderByDescending(x => x.ScheduledAt)
            .Select(x => new CandidateProgressInterviewDto
            {
                InterviewId = x.Id,
                Title = x.Title,
                ScheduledAt = x.ScheduledAt,

                Score = x.Feedback != null
                    ? x.Feedback.OverallScore
                    : null,

                Outcome = x.Feedback != null
                    ? x.Feedback.Outcome
                    : null
            })
            .ToListAsync(cancellationToken);

        var completedInterviews = await _dbContext.Interviews
            .AsNoTracking()
            .CountAsync(
                x => x.CandidateId == candidateId &&
                     x.Status == InterviewStatus.Completed,
                cancellationToken);

        var scores = interviews
            .Where(x => x.Score.HasValue)
            .Select(x => x.Score!.Value)
            .ToList();

        double? averageScore = scores.Count > 0
            ? scores.Average()
            : null;

        var latestOutcome = interviews
            .Where(x => x.Outcome.HasValue)
            .Select(x => x.Outcome)
            .FirstOrDefault();

        return new CandidateProgressDto
        {
            CandidateId = candidate.Id,
            CandidateName =
                $"{candidate.FirstName} {candidate.LastName}",

            TotalInterviews = interviews.Count,
            CompletedInterviews = completedInterviews,
            AverageScore = averageScore,
            LatestOutcome = latestOutcome,
            Interviews = interviews
        };
    }
}

File: frontend/src/app/features/interviewer/reports/services/report.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface ReportOverview {
  totalCandidates: number;
  totalInterviews: number;
  completedInterviews: number;
  scheduledInterviews: number;
  averageScore: number | null;
}

export interface CandidateProgressInterview {
  interviewId: string;
  title: string;
  scheduledAt: string;
  score: number | null;
  outcome: number | null;
}

export interface CandidateProgress {
  candidateId: string;
  candidateName: string;
  totalInterviews: number;
  completedInterviews: number;
  averageScore: number | null;
  latestOutcome: number | null;
  interviews: CandidateProgressInterview[];
}

@Injectable({
  providedIn: 'root'
})
export class ReportService {

  private readonly apiUrl =
    '/api/Reports';

  constructor(
    private readonly http: HttpClient
  ) {}

  getOverview(): Observable<ReportOverview> {
    return this.http.get<ReportOverview>(
      `${this.apiUrl}/overview`,
      {
        withCredentials: true
      }
    );
  }

  getCandidateProgress(
    candidateId: string
  ): Observable<CandidateProgress> {
    return this.http.get<CandidateProgress>(
      `${this.apiUrl}/candidates/${candidateId}/progress`,
      {
        withCredentials: true
      }
    );
  }
}

